src/stan_math_signatures/Stan_math_signatures.ml

```diff
@@ -0,0 +1,159 @@
+  | Greater -> ["logical_gt"]
+  | Geq -> ["logical_gte"]
+  | PNot -> ["logical_negation"]
+  | Transpose -> ["transpose"]
+
+let int_divide_type =
+  UnsizedType.
+    ( [(AutoDiffable, UInt); (AutoDiffable, UInt)]
+    , ReturnType UInt
+    , Fun_kind.FnPlain
+    , Mem_pattern.AoS )
+
+let get_sigs name =
+  let name = Utils.stdlib_distribution_name name in
+  Hashtbl.find_multi stan_math_signatures name
+  |> List.sort ~compare:UnsizedType.compare_signature
+
+let make_assignmentoperator_stan_math_signatures assop =
+  (match assop with
+  | Operator.Divide -> ["divide"]
+  | assop -> operator_to_stan_math_fns assop)
+  |> List.concat_map ~f:get_sigs
+  |> List.concat_map ~f:(function
+       | [(ad1, lhs); (ad2, rhs)], ReturnType rtype, _, _
+         when rtype = lhs
+              && not
+                   ((assop = Operator.EltTimes || assop = Operator.EltDivide)
+                   && UnsizedType.is_scalar_type rtype) ->
+           if rhs = UReal then
+             [ ( [(ad1, lhs); (ad2, UInt)]
+               , UnsizedType.Void
+               , Fun_kind.FnPlain
+               , Mem_pattern.SoA )
+             ; ([(ad1, lhs); (ad2, UReal)], Void, FnPlain, SoA) ]
+           else [([(ad1, lhs); (ad2, rhs)], Void, FnPlain, SoA)]
+       | _ -> [])
+
+let pp_math_sigs ppf name =
+  (Fmt.list ~sep:Fmt.cut UnsizedType.pp_math_sig) ppf (get_sigs name)
+
+let pretty_print_math_sigs = Fmt.str "@[<v>@,%a@]" pp_math_sigs
+
+let string_operator_to_stan_math_fns str =
+  match str with
+  | "Plus__" -> "add"
+  | "PPlus__" -> "plus"
+  | "Minus__" -> "subtract"
+  | "PMinus__" -> "minus"
+  | "Times__" -> "multiply"
+  | "Divide__" -> "divide"
+  | "Modulo__" -> "modulus"
+  | "IntDivide__" -> "divide"
+  | "LDivide__" -> "mdivide_left"
+  | "EltTimes__" -> "elt_multiply"
+  | "EltDivide__" -> "elt_divide"
+  | "Pow__" -> "pow"
+  | "EltPow__" -> "pow"
+  | "Or__" -> "logical_or"
+  | "And__" -> "logical_and"
+  | "Equals__" -> "logical_eq"
+  | "NEquals__" -> "logical_neq"
+  | "Less__" -> "logical_lt"
+  | "Leq__" -> "logical_lte"
+  | "Greater__" -> "logical_gt"
+  | "Geq__" -> "logical_gte"
+  | "PNot__" -> "logical_negation"
+  | "Transpose__" -> "transpose"
+  | _ -> str
+
+let pretty_print_all_math_sigs ppf () =
+  let open Fmt in
+  Format.pp_set_margin ppf 180;
+  let pp_sig ppf (name, (args, rt, _, _)) =
+    pf ppf "%s(@[<h>%a@]) => %a" name
+      (list ~sep:comma UnsizedType.pp)
+      (List.map ~f:snd args) UnsizedType.pp_returntype rt in
+  let pp_sigs_for_name ppf name =
+    (list ~sep:cut pp_sig) ppf
+      (List.map ~f:(fun t -> (name, t)) (get_sigs name)) in
+  pf ppf "@[<v>%a@]"
+    (list ~sep:cut pp_sigs_for_name)
+    (List.sort ~compare:String.compare (Hashtbl.keys stan_math_signatures))
+
+let pretty_print_all_math_distributions ppf () =
+  let open Fmt in
+  let distributions =
+    String.Map.of_alist_reduce distributions ~f:(fun v1 v2 ->
+        v1 @ v2 |> Set.Poly.of_list |> Set.to_list)
+    |> Map.to_alist in
```

Comment:
All this does is dedup `wiener_lpdf`. I think that should be done before marshalling.